Code/Tools/Libs/ToolsFoundation/FileSystem/Implementation/FileSystemModel.cpp

```diff
@@ -0,0 +1,855 @@
+  if (ezPathUtils::IsAbsolutePath(sPath))
+  {
+    it = m_ReferencedFiles.Find(sPath);
+  }
+  else
+  {
+    // Data dir parent relative?
+    for (const auto& dd : m_FileSystemConfig.m_DataDirs)
+    {
+      ezStringBuilder sDataDir;
+      ezFileSystem::ResolveSpecialDirectory(dd.m_sDataDirSpecialPath, sDataDir).IgnoreResult();
+      sDataDir.PathParentDirectory();
+      sDataDir.AppendPath(sPath);
+      it = m_ReferencedFiles.Find(sDataDir);
+      if (it.IsValid())
+        break;
+    }
+
+    if (!it.IsValid())
+    {
+      // Data dir relative?
+      for (const auto& dd : m_FileSystemConfig.m_DataDirs)
+      {
+        ezStringBuilder sDataDir;
+        ezFileSystem::ResolveSpecialDirectory(dd.m_sDataDirSpecialPath, sDataDir).IgnoreResult();
+        sDataDir.AppendPath(sPath);
+        it = m_ReferencedFiles.Find(sDataDir);
+        if (it.IsValid())
+          break;
+      }
+    }
+  }
+
+  if (it.IsValid())
+  {
+    stat = it.Value();
+    return EZ_SUCCESS;
+  }
+  return EZ_FAILURE;
+}
+
+
+ezResult ezFileSystemModel::FindFile(ezDelegate<bool(const ezString&, const ezFileStatus&)> visitor) const
+{
+  if (!m_bInitialized)
+    return EZ_FAILURE;
+
+  EZ_LOCK(m_FilesMutex);
+  for (auto it = m_ReferencedFiles.GetIterator(); it.IsValid(); ++it)
+  {
+    if (visitor(it.Key(), it.Value()))
+      return EZ_SUCCESS;
+  }
+  return EZ_FAILURE;
+}
+
+
+ezResult ezFileSystemModel::LinkDocument(ezStringView sAbsolutePath, const ezUuid& documentId)
+{
+  if (!m_bInitialized)
+    return EZ_FAILURE;
+
+  ezFileStatus fileStatus;
+  bool bDocumentLinkChanged = false;
+  {
+    EZ_LOCK(m_FilesMutex);
+    auto it = m_ReferencedFiles.Find(sAbsolutePath);
+    if (it.IsValid())
+    {
+      bDocumentLinkChanged = it.Value().m_DocumentID != documentId;
+      it.Value().m_DocumentID = documentId;
+      fileStatus = it.Value();
+    }
+    else
+    {
+      return EZ_FAILURE;
+    }
+  }
+
+  if (bDocumentLinkChanged)
+  {
+    FireFileChangedEvent(sAbsolutePath, fileStatus, ezFileChangedEvent::Type::DocumentLinked);
+  }
+  return EZ_SUCCESS;
+}
+
+ezResult ezFileSystemModel::UnlinkDocument(ezStringView sAbsolutePath)
+{
+  if (!m_bInitialized)
+    return EZ_FAILURE;
+
+  ezFileStatus fileStatus;
+  bool bDocumentLinkChanged = false;
+  {
+    EZ_LOCK(m_FilesMutex);
+    auto it = m_ReferencedFiles.Find(sAbsolutePath);
+    if (it.IsValid())
+    {
+      bDocumentLinkChanged = it.Value().m_DocumentID != ezUuid();
+      it.Value().m_DocumentID.SetInvalid();
+      fileStatus = it.Value();
+    }
+    else
+    {
+      return EZ_FAILURE;
+    }
+  }
+
+  if (bDocumentLinkChanged)
+  {
+    FireFileChangedEvent(sAbsolutePath, fileStatus, ezFileChangedEvent::Type::DocumentLinked);
+  }
+  return EZ_SUCCESS;
+}
+
+ezResult ezFileSystemModel::HashFile(ezStringView sAbsolutePath, ezFileStatus& out_stat)
+{
+  if (!m_bInitialized)
+    return EZ_FAILURE;
+
+  // #TODO_ASSET Sanity check that file path is under m_FileSystemConfig
```

Comment:
Actually this as still missing. As this is a public function it needs to check against files outside its data dirs. Otherwise it would call the internal HandleSingleFile function which is not guarded against files outside.

Reply:
Why do we need those files to be hashed?... Oh is it for the Visual shader stuff?

Reply:
Yes visual shader stuff. It makes sense to be hashed but I don't know if we want to expose all of that as a data directory so for now I simply allow hashing files outside data dirs.